### PR TITLE
Add main notebook and supporting modules.

### PR DESCRIPTION
**Pull request by david-stan**

### CODE REVIEW

File: modules/dataset_tokenizer.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+def _dataset_to_features(dataset, tokenizer, set_type):
+    """ Pre-process text. This method will:
+    1) tokenize inputs
+    2) cut or pad each sequence to MAX_SEQ_LENGHT
+    3) convert tokens into ids
+
+    The output will contain:
+    `input_ids` - padded token ids sequence
+    `attention mask` - mask indicating padded tokens
+    `token_type_ids` - mask indicating the split between premise and hypothesis
```
Comment (evgri243) [inline modules/dataset_tokenizer.py]: All good. Just curious, "premise and hypothesis"?.. What?)

File: modules/train.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+            local_model.train()
+            # local dataloader
+            local_train_loader = DataLoader(
+                client_train_partitions[client_id],
+                sampler=RandomSampler(client_train_partitions[client_id]),
+                batch_size=config['BATCH_SIZE'],
+            )
+
+            learning_rate = 0.2
+            if vanilla_lora:
```
Comment (evgri243) [inline modules/train.py]: It's a question of taste, but I'd pass learning rate as a parameter as:
1. It is a quite reasonable parameter
2. it is the only use for vanila_lora

File: modules/train.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+from peft import (
+    get_peft_model_state_dict,
+    set_peft_model_state_dict,
+)
+from tqdm import tqdm
+
+from modules.utils import evaluate
+
+
+def FedAvg(w):
```
Comment (evgri243) [inline modules/train.py]: Is it intentionally made looking a class name? I mean PascalCase instead of snake_case?

File: modules/train.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    w_avg = copy.deepcopy(w[0])
+    for k in w_avg.keys():
+        for i in range(1, len(w)):
+            w_avg[k] += w[i][k]
+        w_avg[k] = torch.div(w_avg[k], len(w))
+    return w_avg
+
+def federated_train(
```
Comment (evgri243) [inline modules/train.py:24]: A commenter died on the second file :)

I also never write comments if not suggested by Copilot

File: modules/utils.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+# --- PARTITIONING ---
+
+def partition_dataset_homogeneous(dataset: TensorDataset, num_clients=3):
+    indices = np.random.permutation(len(dataset))
+    splits = np.array_split(indices, num_clients)
+    return [Subset(dataset, split) for split in splits]
+
+def partition_dataset_heterogeneous(dataset: TensorDataset, num_clients=3):
+    """
```
Comment (evgri243) [inline modules/utils.py]: Sorry, PEP specifies that the docstring should start from the first line)

File: modules/utils.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    return avg_loss, avg_accuracy
+
+# --- PARTITIONING ---
+
+def partition_dataset_homogeneous(dataset: TensorDataset, num_clients=3):
+    indices = np.random.permutation(len(dataset))
+    splits = np.array_split(indices, num_clients)
+    return [Subset(dataset, split) for split in splits]
+
+def partition_dataset_heterogeneous(dataset: TensorDataset, num_clients=3):
```
Comment (evgri243) [inline modules/utils.py]: What's the point of this parameter if proportions have only three elements anyway? Maybe pass proportions and do `for client in range(len(proportions))`?

Honestly, just being annoying...

File: notebook/dp-cls-fed_ffa_lora_main.ipynb
Comment (evgri243) [inline notebook/dp-cls-fed_ffa_lora_main.ipynb:1]: I can't comment on specific lines, but for clarity, may we split trainings of vanilla and FF-LoRa into separate cells?

File: modules/train.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+
+                                print(
+                                    f"      Step: {step} | "
+                                    f"Train loss: {train_loss:.3f} | "
+                                    f"Eval loss: {eval_loss:.3f} | "
+                                    f"Eval accuracy: {eval_accuracy:.3f} | "
+                                    f"ɛ: {eps:.2f}"
+                                )
+
+            print(f"    Local training loss: {np.mean(local_losses):.3f}")
```
Comment (evgri243) [inline modules/train.py:107]: Can we print epsilon in this line just once per client-epoch?

File: notebook/dp-cls-fed_ffa_lora_main.ipynb
Comment (evgri243) [inline notebook/dp-cls-fed_ffa_lora_main.ipynb:1]: Not required now. Just for me, may you also plot a graph per model with accuracy/loss on the 1st y axis and epsilon on the 2nd? Just curious to see how budget is used during both trainings

Comment (evgri243) [reply notebook/dp-cls-fed_ffa_lora_main.ipynb:1]: If we just print these three graphs for models independently, it won't be that clear, right?

Comment (evgri243) [reply notebook/dp-cls-fed_ffa_lora_main.ipynb:1]: Different scales: https://matplotlib.org/stable/gallery/subplots_axes_and_figures/two_scales.html

**Review by evgri243 — APPROVED** [review]

**Comment by evgri243** [comment]
Despite me being generally annoying, brilliant!

File: modules/__init__.py
Comment (evgri243) [inline modules/__init__.py:1]: "modules" make me feel sad but I have nothing better to offer)

File: modules/dataset_tokenizer.py (hunk 1 above)
Comment (david-stan) [reply modules/dataset_tokenizer.py]: They're just names of the columns for this particular dataset. Nevertheless `token_type_ids` are not needed anymore because of the roberta tokenizer.

File: modules/train.py (hunk 2 above)
Comment (david-stan) [reply modules/train.py]: Done.

File: modules/train.py (hunk 3 above)
Comment (david-stan) [reply modules/train.py]: Done.
